File: 97-interleaving-string/97-interleaving-string.py
class Solution:
    def isInterleave(self, s1: str, s2: str, s3: str) -> bool:
        n1, n2, n3 = len(s1), len(s2), len(s3)
        
        if n1 + n2 != n3:
            return False
        
        dp = [[False] * (n2 + 1) for _ in range(n1 + 1)]
        dp[n1][n2] = True
        for p1 in reversed(range(n1 + 1)):
            for p2 in reversed(range(n2 + 1)):
                p3 = p1 + p2
                
                if p1 == n1 and p2 == n2 and p3 == n3:
                    dp[p1][p2] = True
                else:
                    dp[p1][p2] = False
                    if p1 < n1 and s1[p1] == s3[p3]:
                        dp[p1][p2] |= dp[p1 + 1][p2]

                    if p2 < n2 and s2[p2] == s3[p3]:
                        dp[p1][p2] |= dp[p1][p2 + 1]
        return dp[0][0]
        

File: 97-interleaving-string/97-interleaving-string.java
class Solution {
    public boolean isInterleave(String s1, String s2, String s3) {
        if(s1.length() + s2.length() != s3.length())
            return false;
        
        boolean[] prev = new boolean[s2.length() + 1];
        boolean[] curr = new boolean[s2.length() + 1];
        for(int i = s1.length(); i >= 0; i--){
            for(int j = s2.length(); j >= 0; j--){
                int k = i + j;
                
                if(k == s3.length()){
                    curr[j] = true;
                    continue;
                }
                
                curr[j] = false;
                if(i < s1.length() && s1.charAt(i) == s3.charAt(k))
                    curr[j] = prev[j];
                
                if(j < s2.length() && s2.charAt(j) == s3.charAt(k))
                    curr[j] = curr[j] || curr[j + 1];
            }
            boolean[] tmp = prev;
            prev = curr;
            curr = tmp;
        }
        return prev[0];
    }
}